text/0068-feat-login-into-portable-account-with-metamask.md

```diff
@@ -0,0 +1,116 @@
+
+# Context, motivation and guide level explanation
+
+The solution worked out here allows to log in to the user account from any browser into the same account described in [FIP-59](https://github.com/fairDataSociety/FIPs/blob/master/text/0059-portable-account.md) created with
+[fdp-create-account](https://github.com/fairDataSociety/fdp-create-account) project.
+
+# Reference-level explanation
+
+For the login to work with metamask, user should already have a portable account created with [fdp-create-account](https://github.com/fairDataSociety/fdp-create-account).
+
+As we know while creating a portable account users can provide a 12 word mnemonic phrase for the portable wallet or a 12 word mnemonic will be generated for them.
+
+This solution starts with importing the portable wallet into metamask. After that it is a two-step process
+
+## Step 1: 
+```go
+ConnectPortableAccountWithWallet(userName, passPhrase, addressHex, signature string) error
+```
+
+This function will get the portable wallet (wallet seed) and re-upload an alternate socTopic described later
+
+## Step 2: 
+```go
+LoginWithWallet(addressHex, signature string) (*user.Info, error) 
+```
+
+This function will get the encrypted portable wallet (wallet seed) and username from the alternate socTopic and decrypt it with the signature provided by the user.
+
+
+## Uploading portable wallet (wallet seed) to Ethereum Swarm to an alternate socTopic
+
+In FIP-59 we have discussed, about the socTopic which is a topic in the swarm where the portable account is stored. In this solution we will be using an alternate socTopic, but with the same logic, to store the wallet seed.
+```
+socTopic = H(fdpLoginVersion + portableWalletAddress + signature)
+```
+
+Where `portableWalletAddress` and the `signature` will be provided by the user. As we have already imported the portable wallet into metamask, we can get the `portableWalletAddress` from there. The `signature` will be generated by signing the `portableWalletAddress` with metamask with the portable wallet.
+
+## Wallet seed encryption
+In FIP-59 we have discussed, about the encryption of wallet seed and storing it in a chunk. 
+
+- We will first generate FIP-59 socTopic with username and password provided by the user.
+- Then download the wallet seed chunk from FIP-59 socTopic and decrypt it with the password provided by the user.
+
+We have the seed now. We will store the username with the seed for identifying the user while logging in with metamask.
+
+### Creating the chunk
+
+The content of the chunk will look something lite this
+
+```
+SEED + USERNAME_SIZE + USERNAME + RANDOM_DATA
+```
+
+We store the username length in 8 bytes, then the username and then the random data.
+
+The `SEED + USERNAME_SIZE + USERNAME` will be padded with random data with byte length `CHUNK_SIZE - SEED_SIZE - IV_LENGTH - USERNAME_LENGTH - USERNAME_SIZE`.
+
+*** `USERNAME_SIZE` is 8 bytes, `USERNAME_LENGTH` is the length of the username provided by the user.

```

Comment:
A bit confusing you use `USERNAME_SIZE` in two different interpretation in one sentence

Reply:
`USERNAME_SIZE` is the byte array size that will store the length of the username. this will be a constant. On the other hand `USERNAME_LENGTH` is the actual length of the username

Reply:
I thought that the `USERNAME_SIZE` indicates the length of `USERNAME`  in `SEED + USERNAME_SIZE + USERNAME`. Otherwise how will you know where the `USERNAME` ends?

Reply:
`USERNAME_SIZE` is the 8 byte long array that saves `USERNAME_LENGTH`. 

So if the username is `Jhon`, `USERNAME_LENGTH` is 4 and `USERNAME_SIZE` is `[4 0 0 0 0 0 0 0]` .

We will make `USERNAME_SIZE` to 1 byte apparently, which will make it less confusing.